pyodide-build/pyodide_build/pyzip.py

```diff
@@ -0,0 +1,128 @@
+#       So we don't need this after we upgrade to 3.12.0
+NOT_ZIPPED_FILES = ("ctypes/", "unittest/")
+
+
+def default_filterfunc(
+    root: Path, verbose: bool = False
+) -> Callable[[str, list[str]], set[str]]:
+    def filterfunc(path: Path | str, names: list[str]) -> set[str]:
+        filtered_files = {
+            (root / f).resolve()
+            for f in REMOVED_FILES + UNVENDORED_FILES + NOT_ZIPPED_FILES
+        }
+
+        path = Path(path).resolve()
+
+        if path.name == "__pycache__":
+            return set(names)
+
+        _names = []
+        for name in names:
+
+            fullpath = path / name
+
+            if fullpath.name == "__pycache__" or fullpath in filtered_files:
+                if verbose:
+                    print(f"Skipping {fullpath}")
+
+                _names.append(name)
+
+        return set(_names)
+
+    return filterfunc
+
+
+def create_zip(
+    libdir: Path | str,
+    output: Path | str = "python",
+    pycompile: bool = False,
+    filterfunc: Callable[[str, list[str]], set[str]] | None = None,
+) -> None:
+    """
+    Bundle Python standard libraries into a zip file.
+
+    The basic idea of this function is similar to the standard library's
+    [PyZipFile](https://github.com/python/cpython/blob/59665d0280c2299ea87e9af45cedc90656cb6f55/Lib/zipfile/__init__.py#L2011).
+
+    However, we need some additional functionality for Pyodide. For example:
+
+    - We need to remove some unvendored modules, e.g. `sqlite3`
+    - We need an option to "not" compile the files in the zip file
+
+    hence this function.
+
+    Parameters
+    ----------
+    libdir : Path or str
+        Path to the directory containing the Python standard library.
+
+    output : Path or str, optional
+        Path to the output zip file. Defaults to python.zip.
+
+    pycompile : bool, optional
+        Whether to compile the .py files into .pyc, by default False
+
+    filterfunc : Callable[[str, list[str]], set[str]], optional
+        A function that filters the files to be included in the zip file.
+        This function will be passed to shutil.copytree's ignore argument.
+        By default, Pyodide's default filter function is used.
+
+    Returns
+    -------
+    BytesIO
+        A BytesIO object containing the zip file.
+    """
+
+    if pycompile:
+        raise NotImplementedError(
+            "TODO: implement after https://github.com/pyodide/pyodide/pull/3253 is merged"
+        )
+
+    libdir = Path(libdir)
+    output = Path(output)
+    output = output.with_name(output.name.rstrip(".zip"))
+
+    if filterfunc is None:
+        filterfunc = default_filterfunc(libdir)
+
+    with TemporaryDirectory() as temp_dir_str:
+        temp_dir = Path(temp_dir_str)
+        shutil.copytree(libdir, temp_dir, ignore=filterfunc, dirs_exist_ok=True)
+
+        shutil.make_archive(str(output), "zip", temp_dir)
```

Comment:
Does it use compression by default?

Reply:
Yeah it does, though I think it is not a very well documented behavior. (https://github.com/python/cpython/blob/31b639a99281bd9582e88137a3be2868c4154a88/Lib/shutil.py#L978)